src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1558,3 +2303,391 @@ MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
+
+	if (node == NULL)
+	{
+		return NULL;
+	}
+
+	/*
+	 * Look for operator expressions with two arguments.
+	 *
+	 * Once Found the  uninstantiate, replace with appropriate boundaries for the
+	 * current shard interval.
+	 *
+	 * The boundaries are replaced in the following manner:
+	 * (partCol >= shardMinValue && partCol <= shardMaxValue)
+	 */
+	if (IsA(node, OpExpr) && list_length(((OpExpr *) node)->args) == 2)
+	{
+		OpExpr *op = (OpExpr *) node;
+		Node *leftop = get_leftop((Expr *) op);
+		Node *rightop = get_rightop((Expr *) op);
+		Param *param = NULL;
+
+		Var *hashedGEColumn = NULL;
+		OpExpr *hashedGEOpExpr = NULL;
+		Datum shardMinValue = shardInterval->minValue;
+
+		Var *hashedLEColumn = NULL;
+		OpExpr *hashedLEOpExpr = NULL;
+		Datum shardMaxValue = shardInterval->maxValue;
+
+		List *hashedOperatorList = NIL;
+
+		Oid integer4GEoperatorId = InvalidOid;
+		Oid integer4LEoperatorId = InvalidOid;
+
+		/* look for the Params */
+		if (IsA(leftop, Param))
+		{
+			param = (Param *) leftop;
+		}
+		else if (IsA(rightop, Param))
+		{
+			param = (Param *) rightop;
+		}
+
+		/* not an interesting param for our purpose, so return */
+		if (!(param && param->paramid == UNINSTANTIATED_PARAMETER_ID))
+		{
+			return node;
+		}
+
+		/* get the integer >=, <= operators from the catalog */
+		integer4GEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+												   INT4OID,
+												   BTGreaterEqualStrategyNumber);
+		integer4LEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+												   INT4OID,
+												   BTLessEqualStrategyNumber);
+
+		/* generate hashed columns */
+		hashedGEColumn = MakeInt4Column();
+		hashedLEColumn = MakeInt4Column();
+
+		/* generate the necessary operators */
+		hashedGEOpExpr = (OpExpr *) make_opclause(integer4GEoperatorId,
+												  InvalidOid, false,
+												  (Expr *) hashedGEColumn,
+												  (Expr *) MakeInt4Constant(
+													  shardMinValue),
+												  InvalidOid, InvalidOid);
+
+		hashedLEOpExpr = (OpExpr *) make_opclause(integer4LEoperatorId,
+												  InvalidOid, false,
+												  (Expr *) hashedLEColumn,
+												  (Expr *) MakeInt4Constant(
+													  shardMaxValue),
+												  InvalidOid, InvalidOid);
+
+		/* update the operators with correct operator numbers and function ids */
+		hashedGEOpExpr->opfuncid = get_opcode(hashedGEOpExpr->opno);
+		hashedGEOpExpr->opresulttype = get_func_rettype(hashedGEOpExpr->opfuncid);
+
+		hashedLEOpExpr->opfuncid = get_opcode(hashedLEOpExpr->opno);
+		hashedLEOpExpr->opresulttype = get_func_rettype(hashedLEOpExpr->opfuncid);
+
+		/* finally add the hashed operators to a list and return it */
+		hashedOperatorList = lappend(hashedOperatorList, hashedGEOpExpr);
+		hashedOperatorList = lappend(hashedOperatorList, hashedLEOpExpr);
+
+		return (Node *) hashedOperatorList;
+	}
+
+
+	if (IsA(node, Query))
+	{
+		/* FIXME: probably can remove support for this */
```

Comment:
Yes, I think so. I even think that we do not need to Assert as well, but, keeping it for now.